File: scala/src/main/scala/Day5.scala
// enum might take parameteres here
enum Car:
  case FastCar(maxSpeed: Int)
  case OldCar(age: Int)

// this will emit a warning since the compiler is able to notice that
// there are cases on which this function will return exception
def testCar(car: Car): Unit = car match {
  case Car.FastCar(maxSpeed) =>
    println(s"maximal speed of fast car is $maxSpeed")
}

@main def enums(): Unit =
  val myCar = Car.FastCar(100)
  val ordinal = myCar.ordinal

  println(s"my fast car ordinal is $ordinal")

File: rust/examples/day5.rs
enum Car {
    FastCar { speed: usize },
    _OldCar { age: u8 },
}

fn show_car(car: &Car) {
    match car {
        Car::FastCar { speed } => println!("my fast car can ride up to {}", speed),
        // this results in warning only in scala
        _ => unimplemented!(),
    }
}
fn main() {
    let my_car = Car::FastCar { speed: 100 };
    show_car(&my_car);
    let index = std::mem::discriminant(&my_car);
    // this will print:
    // my fast car has index of Discriminant(0)
    println!("my fast car has index of {index:?}")
}
